### PR TITLE
Draft Test

### PR DESCRIPTION
**Pull request by HoshinoTented**
I added some tests for `f32_add`, but I don't know whether they work correctly.

### CODE REVIEW

File: src/soft_f32/soft_f32_add.rs [hunk 1]
```diff
@@ -165,14 +168,17 @@ mod tests {
...
 
+        // common tests about NaN
+        crate::soft_f32::test_util::test_nan(super::f32_add);
+
         // NaN + 1 = NaN
         assert_eq!(crate::soft_f32::f32_add(0xFFFFFFFF, 0x3F800000), 0xFFFFFFFF);
 
         // NaN + -1 = NaN
-        assert_eq!(crate::soft_f32::f32_is_nan(crate::soft_f32::f32_add(0xFFFFFFFF, 0x3F800000)), true);
+        assert_eq!(crate::soft_f32::f32_is_nan(crate::soft_f32::f32_add(0xFFFFFFFF, 0x3F800000)), true);        // what is it? 1 == -1?
```
Comment (Inokinoki) [inline src/soft_f32/soft_f32_add.rs]: NaN does not conform reflexion. So, we cannot, from `NaN == NaN`, imply `NaN + 1 == NaN + -1` and thus `1 == -1`.

I find some resources which may help you understand this in a more rustic way: by making differences between [PartialEq](https://doc.rust-lang.org/std/cmp/trait.PartialEq.html) and [Eq](https://doc.rust-lang.org/std/cmp/trait.Eq.html).

File: src/soft_f32/soft_f32_add.rs [hunk 2]
```diff
@@ -165,14 +168,17 @@ mod tests {
...
 
         // NaN + Inf = NaN
-        assert_eq!(crate::soft_f32::f32_is_nan(crate::soft_f32::f32_add(0xFFFFFFFF, 0x7F800000)), true);
+        assert_eq!(crate::soft_f32::f32_is_nan(crate::soft_f32::f32_add(0xFFFFFFFF, 0x7F800000)), true);        // why not use assert!()
```
Comment (Inokinoki) [inline src/soft_f32/soft_f32_add.rs]: Some copy-paste dirty work.

**Comment by Inokinoki** [comment]
I'm not declining more contributions, but you can only add some tests for multiplication and division instead of all the functions. Just do not want to bring you too much work.

But go ahead if you'd like. I hope you can learn sth from this project and teach me sth in Rust.

**Comment by HoshinoTented** [comment]
Oh no :(
I made this draft as a pull request.

File: src/soft_f32/test_util.rs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+pub(crate) const ONE: u32 = 0x3F800000;
+pub(crate) const NEG_ONE: u32 = 0xBF800000;
+
+pub fn test_nan<F>(f: F)
+    where
+        F: Fn(u32, u32) -> u32 {
+
+    // NaN `f` ± 1 = NaN
+    assert!(f32_is_nan(f(0xFFFFFFFF, ONE)));
+    assert!(f32_is_nan(f(0xFFFFFFFF, NEG_ONE)));
```
Comment (HoshinoTented) [inline src/soft_f32/test_util.rs]: So, should I use NEG_ONE here?

File: src/soft_f32/soft_f32_sub.rs [hunk 4]
```diff
@@ -128,4 +132,33 @@ mod tests {
...
+        // Inf - ±1 = Inf
+        assert_eq!(f32_sub(INF, ONE), INF);
+        assert_eq!(f32_sub(INF, NEG_ONE), INF);
+
+        // -Inf - ±1 = -Inf
+        assert_eq!(f32_sub(NEG_INF, ONE), NEG_INF);
+        assert_eq!(f32_sub(NEG_INF, NEG_ONE), NEG_INF);
+
+        // ±Inf - ±Inf = NaN
+        assert!(f32_is_nan(f32_sub(INF, INF)));
```
Comment (HoshinoTented) [inline src/soft_f32/soft_f32_sub.rs:153]: assertion failed in this line.

**Pull request closed without merging**